core/node/api_server/src/utils.rs

```diff
@@ -66,3 +77,383 @@ macro_rules! report_filter {
+        let mut storage = pool.connection().await.unwrap();
+        prepare_storage(&mut storage, alice.address()).await;
+        persist_block_with_transactions(&pool, txs).await;
+
+        // Sanity check: for successful deployments, the actual deployed address must correspond to the derived one.
+        let deployment_events = storage
+            .events_web3_dal()
+            .get_contract_deployment_logs(L2BlockNumber(1))
+            .await
+            .unwrap();
+        let expected_events: Vec<_> = expected_statuses_and_contract_addresses
+            .iter()
+            .enumerate()
+            .filter_map(|(i, &(status, address))| {
+                if let (1, Some(deployed_address)) = (status, address) {
+                    Some(ContractDeploymentLog {
+                        transaction_index_in_block: i as u64,
+                        deployer: alice.address(),
+                        deployed_address,
+                    })
+                } else {
+                    None
+                }
+            })
+            .collect();
+        assert_eq!(deployment_events, expected_events);
+
+        for (&tx_hash, &(status, expected_address)) in tx_hashes
+            .iter()
+            .zip(&expected_statuses_and_contract_addresses)
+        {
+            println!("Fetching receipt for {tx_hash:?}");
+            let receipts = storage
+                .transactions_web3_dal()
+                .get_transaction_receipts(&[tx_hash])
+                .await
+                .unwrap();
+            let filled_receipts = fill_transaction_receipts(&mut storage, receipts)
+                .await
+                .unwrap();
+            assert_eq!(filled_receipts.len(), 1);
+            let receipt = filled_receipts.into_iter().next().unwrap();
+            assert_eq!(receipt.status, status.into());
+            assert_eq!(receipt.contract_address, expected_address);
+        }
+
+        // Test all receipts for a block.
+        let receipts = storage
+            .transactions_web3_dal()
+            .get_transaction_receipts(&tx_hashes)
+            .await
+            .unwrap();
+        let filled_receipts = fill_transaction_receipts(&mut storage, receipts)
+            .await
+            .unwrap();
+        let statuses_and_contract_addresses: Vec<_> = filled_receipts
+            .iter()
+            .map(|receipt| (receipt.status.as_u32(), receipt.contract_address))
+            .collect();
+        assert_eq!(
+            statuses_and_contract_addresses,
+            expected_statuses_and_contract_addresses
+        );
+    }
+
+    /// Because of AA support, determining deployment nonces is non-trivial. E.g., it would be incorrect
+    /// to define a deployment nonce increment as a count of successful deployment txs from an EOA because
+    /// the account might have a non-default AA that can be called as a contract (so it can deploy contracts
+    /// outside of deployment txs).
```

Comment:
Just to understand better: Does this logic apply in practice (i.e., there are no implied limitations of what AA contracts can do)?

Reply:
You are right.